src/ray/gcs/gcs_server/gcs_task_manager.h

```diff
@@ -203,18 +237,136 @@ class GcsTaskManager : public rpc::TaskInfoHandler {
+    /// Each locator contains the iterator to the list and the index of the list.
+    /// - When a task event is added to the storage, a locator is created and added to the
+    /// indices.
+    /// - When a task event is removed from the storage, the locator is removed from the
+    /// indices.
+    /// - When a task event is updated, it might move between different lists, and the
+    /// locator will be updated accordingly.
+    class TaskEventLocator {
+     public:
+      TaskEventLocator(std::list<rpc::TaskEvents>::iterator iter, size_t task_list_index)
+          : iter_(iter), task_list_index_(task_list_index) {}
+
+      rpc::TaskEvents &GetTaskEvents() const { return *iter_; }
+
+      size_t GetCurrentListIndex() const { return task_list_index_; }
+
+      std::list<rpc::TaskEvents>::iterator GetCurrentListIterator() const {
+        return iter_;
+      }
+
+      void SetCurrentList(size_t cur_list_index,
+                          std::list<rpc::TaskEvents>::iterator cur_list_iter) {
+        iter_ = cur_list_iter;
+        task_list_index_ = cur_list_index;
+      }
+
+     private:
+      /// Iterator to the task list.
+      std::list<rpc::TaskEvents>::iterator iter_;
+      /// Index of the task list.
+      size_t task_list_index_;
+    };
+
+    /// A helper class to summarize the stats of a job.
+    /// TODO: we could probably do source side summary here per job.
+    ///
+    /// This class contains stats of:
+    /// - Number of task attempts dropped, it's used to determine if task events should be
+    /// dropped if data from the task attempt is being already dropped.
+    /// - Number of profile events dropped.
+    class JobTaskSummary {
+     public:
+      /// Record a task attempt as dropped.
+      ///
+      /// \param task_attempt Task attempt.
+      void RecordTaskAttemptDropped(const TaskAttempt &task_attempt) {
+        auto inserted = dropped_task_attempts_.insert(task_attempt);
+        if (inserted.second) {
+          num_task_attempts_dropped_++;
+        }
+      }
+
+      /// Record a number of profile event as dropped.
+      void RecordProfileEventsDropped(int32_t cnt) { num_profile_events_dropped_ += cnt; }
+
+      /// Return if a task attempt should be dropped.
+      ///
+      /// A task attempt should be dropped if some task events from the attempt are
+      /// already dropped.
+      bool ShouldDropTaskAttempt(const TaskAttempt &task_attempt) const {
+        return dropped_task_attempts_.count(task_attempt) > 0;
+      }
+
+      size_t NumProfileEventsDropped() const { return num_profile_events_dropped_; }
+
+      size_t NumTaskAttemptsDropped() const { return num_task_attempts_dropped_; }
+
+      /// Callback when job is finished.
+      ///
+      /// When a job is finished, there will be no more task events from the job. So we
+      /// can clear the cached dropped task attempts.
+      void OnJobEnds() { dropped_task_attempts_.clear(); }
```

Comment:
can you add unit tests that verify this clean up?

Reply:
added `TestMultipleJobsDataLoss`